src/manager/node.c

```diff
@@ -493,32 +543,78 @@ static int node_method_start_unit(sd_bus_message *m, void *userdata, UNUSED sd_b
-        if (op == NULL) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
+        req = node_create_request(
+                        node,
+                        "StartUnit",
+                        method_start_unit_callback,
+                        job_setup_ref(setup),
+                        (free_func_t) job_setup_unref);
+        if (req == NULL) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
         }
 
-        /* TODO: Handle mode, queueing job as needed */
+        sd_bus_message_append(req->message, "ssu", unit, mode, setup->job->id);
 
-        op->request_message = sd_bus_message_ref(m);
-        op->job = job_new(node, unit, "Start");
-        if (op->job == NULL) {
+        if (!agent_request_start(req)) {
                 return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
         }
 
-        _cleanup_agent_request_ AgentRequest *req = node_create_request(
-                        node, "StartUnit", method_start_unit_callback, op, (free_func_t) start_unit_op_free);
-        if (req == NULL) {
+        return 1;
+}
+
+/*************************************************************************
+ ********** org.containers.hirte.Node.StopUnit **************************
+ ************************************************************************/
+
+static int method_stop_unit_callback(AgentRequest *req, UNUSED sd_bus_message *m, UNUSED sd_bus_error *ret_error) {
+        Node *node = req->node;
+        Manager *manager = node->manager;
+        JobSetup *setup = req->userdata;
+
+        if (sd_bus_message_is_method_error(m, NULL)) {
+                /* Forward error */
+                return sd_bus_reply_method_error(setup->request_message, sd_bus_message_get_error(m));
+        }
+
+        if (!manager_add_job(manager, setup->job)) {
                 return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
         }
 
-        sd_bus_message_append(req->message, "ssu", unit, mode, op->job->id);
+        return sd_bus_reply_method_return(setup->request_message, "o", setup->job->object_path);
+}
+
+static int node_method_stop_unit(sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        Node *node = userdata;
+        _cleanup_agent_request_ AgentRequest *req = NULL;
+        const char *unit = NULL;
+        const char *mode = NULL;
+
+        int r = sd_bus_message_read(m, "ss", &unit, &mode);
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
+        _cleanup_job_setup_ JobSetup *setup = job_setup_new(m, node, unit, "Stop");
+        if (setup == NULL) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
+        }
+
+        req = node_create_request(
+                        node,
+                        "StopUnit",
```

Comment:
Does it make sense to define constants for the method/signal/property names?

Reply:
Yeah, probably. In protocol.h

Reply:
Sounds good. Its just a minor details, though, and doesn't need to be done in this PR.

Reply:
I'll added an issue for this: https://github.com/containers/hirte/issues/109